README.md
# screenshot-service

My main goal in creating this service was to create a solution that allowed me to obtain valuable information from phishing emails that attempted to lure me to a malicious URL. This service will obtian a screenshot, as well as DNS records, associated with the domain.

The benefit of using a service like this was that I would not be making any of these requests from my home network IP, and I can even automate this by caling the API.

# Quick Start
Download [Terraform](https://www.terraform.io/downloads.html) and place the executable in the directory.  
```
git clone --branch v2.0.0 https://github.com/pendraggon87/screenshot-service.git
cd screenshot-service  
wget https://releases.hashicorp.com/terraform/0.12.12/terraform_0.12.12_windows_amd64.zip
unzip terraform_0.12.12_windows_amd64.zip && rm terraform_0.12.12_windows_amd64.zip
```

Configure the `provider.tf` file with your appropriate AWS access keys, or other mechanisms to ensure Terraform is able to operate. Terraform is currently configured to store state in an already existing S3 bucket - this must be created prior to running any Terraform commands. To store state locally, simply remove the Terraform backend declaration.
> There are many ways to handle granting Terraform AWS access. At this stage, the template is using access keys for ease of use

If you do not want to use a custom domain name, or you do not use [Cloudflare](https://cloudflare.com) as your DNS provider, remove the provider information from `provider.tf`.

After you have provided the appropriate credentials, review the `variables.tf` and configure the `terraform.tfvars` file to include any relevant variables.

Run `.\terraform.exe plan`, and if there are no errors, run `.\terraform.exe apply`.  The application will return the API endpoint and the API key for use - that's it!

## Generating those Screenshots
The API Gateway accepts only **GET** requests. If successful, a JSON response will be returned - the "screenshot_url" key will contain the URL to your newly generated image!
### GET Request
Simply pass the url (do NOT URL encode the string) to the gateway using the **url** query parameter, and you are good to go!  
`https://RANDOM.execute-api.us-east-1.amazonaws.com/prod/screenshot?url=http://caliburnsecurity.com`

## Obtaining DNS records
The API Gateway accepts only **GET** requests.  If successful, a JSON response will be returned containing the values for obtained DNS queries. By default, the following records will be queried:
* MX
* SOA
* NS
* SRV
* A
* AAAA
* CNAME
* TXT
### GET Request
Simply pass the FQDN to the gateway using the **FQDN** query parameter, and you are good to go!
`https://RANDOM.execute-api.us-east-1.amazonaws.com/prod/dns?FQDN=caliburnsecurity.com

# Notes
* If a URI scheme is not defined, or the scheme is not one of **https** or **http**, the default scheme selected will be ***http*** unless otherwise configured
* Non-standard ports must be explicitly added to the URL (e.g. *http://github.com:8080*) - otherwise, the default of 80 or 443 will be applied based on the provided scheme
* Only one URL can be passed at this time
* Version of headless chrome (1.0.0-54) found here: https://github.com/adieuadieu/serverless-chrome/releases/download/v1.0.0-54/stable-headless-chromium-amazonlinux-2017-03.zip
* Version of chromedriver used (2.41) can be found here: https://chromedriver.storage.googleapis.com/2.41/chromedriver_linux64.zip

CHANGELOG.md
# CHANGELOG
All notable changes to this project will be documented in this file (as of version 20.09.03).

## 20.09.03

### Added
* Added this changelog 😄
* Added an `/analysis` endpoint to return both a screenshot and DNS entries (simply calls the `/screenshot` and `/dns` endpoints)

lambda/analysis/analysis.py
import requests
import os
import logging
import json
from urllib.parse import urlparse, unquote

# Configure logging
logger = logging.getLogger()
logger.setLevel(logging.DEBUG)

def handler(event, context):
    logger.debug("## ENVIRONMENT VARIABLES ##")
    logger.debug(os.environ)
    logger.debug("## EVENT ##")
    logger.debug(event)

    if event["queryStringParameters"] is None or "url" not in event["queryStringParameters"]:
        resp = {
            "message": "Please submit a valid URL or FQDN"
        }
        return {
            "statusCode": 400,
            "body": json.dumps(resp)
        }

    url = get_url(event["queryStringParameters"]["url"])

    warnings = []
    resp = {}
    headers = {
        "x-api-key": os.environ["apigw_key"]
    }
    logger.info(f"https://api.caliburnsecurity.com/screenshot?url={url.geturl()}")
    r = requests.get(f"https://api.caliburnsecurity.com/screenshot?url={url.geturl()}", headers=headers)
    data = r.json()
    logger.debug(f"SCREENSHOT DEBUG: {data}")

    if r.status_code != 200:
        warnings.append(data["message"])
    else:
        resp["screenshot_url"] = data["screenshot_url"]

    logger.debug(f"FQDN: {url.netloc.split(':')[0]}")

    r = requests.get(f"https://api.caliburnsecurity.com/dns?FQDN={url.netloc.split(':')[0]}", headers=headers)
    data = r.json()
    logger.debug(f"DNS DATA: {data}")

    if r.status_code != 200:
        warnings.append(data["message"])
    else:
        resp["dns_records"] = data["dns_records"]
        if len(data["warnings"]) > 0:
            warnings.extend(data["warnings"])
    
    return {
        "statusCode": 200,
        "body": json.dumps(resp)
    }
    
def get_url(url, https_default=bool(os.environ["https_default"])):
    logger.info(f"Decoding {url}")
    url = unquote(url)

    logger.info(f"Parsing {url}")
    try:
        url = urlparse(url)
        if url.scheme != "http" and url.scheme != "https":
            url = urlparse(f"https://{url}") if https_default == True else urlparse(f"http://{url}")
        if url.port is None:
            url = urlparse(f"{url.geturl()}:443") if https_default == True else urlparse(f"{url.geturl()}:80")
    except Exception as e:
        logger.error(e)
        return {
            "statusCode": "400",
            "message": "Invalid URL provided",
            "error": str(e)
        }
    logger.info(f"URL: {url.geturl()}")
    return url

lambda/dns-service/dns-service.py
import logging
import os
import dns.resolver
import json
from ast import literal_eval
import sys

# Configure logging
logger = logging.getLogger()
logger.setLevel(logging.DEBUG)

def handler(event, context):
    logger.debug("## ENVIRONMENT VARIABLES ##")
    logger.debug(os.environ)
    logger.debug("## EVENT ##")
    logger.debug(event)   


    if event["queryStringParameters"] is None or "FQDN" not in event["queryStringParameters"]:
        resp = {
            "message": "Please submit a valid FQDN"
        }
        return {
            "statusCode": 400,
            "body": json.dumps(resp)
        }

    RESOLVER_TIMEOUT = int(os.environ["RESOLVER_TIMEOUT"])
    RESOLVER_LIFETIME = int(os.environ["RESOLVER_LIFETIME"])
    DNS_RECORD_TYPES = literal_eval(os.environ["DNS_RECORD_TYPES"]) if event["queryStringParameters"] is not None and "recordTypes" not in event["queryStringParameters"] else event["queryStringParameters"]["recordTypes"].split(",") # Not sure why this is necessary, I am passing the array in...


    FQDN = event["queryStringParameters"]["FQDN"]

    logger.info(f"Obtaining DNS records from {FQDN}")
    resolver = dns.resolver.Resolver()
    logger.debug(f"Setting DNS timeout to {RESOLVER_TIMEOUT} seconds")
    resolver.timeout = 1
    logger.debug(f"Setting the query lifetime to {RESOLVER_LIFETIME} seconds")
    resolver.lifetime = 1

    dns_records = {}
    warnings = []

    for record in DNS_RECORD_TYPES:
        logger.debug(f"Record: {record}")
        try:
            answer = resolver.query(FQDN, record)
            for rdata in answer:
                if record.lower() not in dns_records.keys():
                    dns_records[record.lower()] = []
                dns_records[record.lower()].append(rdata.to_text())
        except dns.exception.DNSException as e:
            logger.warning(e)
            warnings.append(str(e))

    resp = {
        "message": f"Successfully performed DNS lookups against {FQDN}",
        "dns_records": dns_records,
        "warnings": warnings
    }

    logger.info(dns_records)

    return {
        "statusCode": 200,
        "body": json.dumps(resp)
    }

lambda/screenshot-service-processor/screenshot-service-processor.py
#!/usr/bin/env python
# -*- coding utf-8 -*-

import json
import logging
import os
import boto3
from datetime import datetime

# Configure logging
logger = logging.getLogger()
logger.setLevel(logging.DEBUG)

def handler(event, context):
    logger.debug("## ENVIRONMENT VARIABLES ##")
    logger.debug(os.environ)
    logger.debug("## EVENT ##")
    logger.debug(event)

    bucket_name = os.environ["s3_bucket"]
    logger.debug(f"bucket_name: {bucket_name}")

    logger.info("Extracting urls")

    urls = []
    if event["httpMethod"] == "GET":
        logger.debug(f"HTTP Method: GET")
        if event["queryStringParameters"]:
            try:
                urls.append(event["queryStringParameters"]["url"])
            except Exception as e:
                logger.error(e)
                return {
                    "statusCode": 500,
                    "body": json.dumps(e)
                }
        else:
            return {
                "statusCode": 400,
                "body": json.dumps("No URL provided...")
            }
    elif event["httpMethod"] == "POST":
        logger.debug("HTTP Method: POST")
        if event["body"]:
            try:
                body = json.loads(event["body"])
                if "urls" in body:
                    logger.info(f"{len(body['urls'])} urls submitted for processing")
                    urls = body["urls"]
                else:
                    urls.append(body["url"])
            except Exception as e:
                logger.error(e)
                return {
                    "statusCode": 500,
                    "body": json.dumps(e)
                }
        else:
            return {
                "statusCode": 400,
                "body": json.dumps("No URL provided...")
            }
    else:
        return {
            "statusCode": 405,
            "body": json.dumps(f"Invalid HTTP Method supplied: {event['httpMethod']}")
        }
    
    bucket_key = datetime.utcnow().strftime('%Y%m%d_%H%M%S')
    logger.info(f"Creating S3 bucket: {bucket_name}/{bucket_key}")

    s3 = boto3.client("s3")
    s3.put_object(Bucket=bucket_name, Key=(f"{bucket_key}/"), ACL="public-read")
    data_test = b"TEST"
    #s3.upload_file(f"/tmp/{screenshot_title}.png", s3_bucket, f"{screenshot_title}.png", ExtraArgs={'ContentType': 'image/png', 'ACL': 'public-read'})
    s3.put_object(Bucket=bucket_name, Key=(f"{bucket_key}/test.txt"), Body=data_test, ACL="public-read", ContentType="text/plain")

    data = {
        "urls": urls,
        "bucket_name": bucket_name,
        "bucket_key": bucket_key
    }
    lambda_cli = boto3.client("lambda")
    lambda_cli.invoke(FunctionName=context.function_name, InvocationType="Event", Payload=json.dumps(data))

    response_body = {
        "message": "Folder successfully created in S3 bucket for URLs. Retrieving screenshots now!",
        "s3_path": f"https://{bucket_name}.s3.amazonaws.com/{bucket_key}/"
    }

    return {
        "statusCode": 200,
        "body": json.dumps(response_body)
    }


lambda/screenshot-service/screenshot-service.py
#!/usr/bin/env python
# -*- coding utf-8 -*-

# TODO: Add logging and cloudwatch
# TODO: Leverage lambda layers so that I can supply the chrome executables without pulling each time

import json
import logging
from urllib.parse import urlparse, unquote # TODO: Can I use urllib3?
from selenium import webdriver
from datetime import datetime
import os
from shutil import copyfile
import boto3
import stat
import urllib.request
import tldextract

# Configure logging
logger = logging.getLogger()
logger.setLevel(logging.DEBUG)

def configure_binaries():
    """Copy the binary files from the lambda layer to /tmp and make them executable"""
    copyfile("/opt/chromedriver", "/tmp/chromedriver")
    copyfile("/opt/headless-chromium", "/tmp/headless-chromium")

    os.chmod("/tmp/chromedriver", 755)
    os.chmod("/tmp/headless-chromium", 755)

def get_screenshot(url, s3_bucket, screenshot_title = None): 
#   TODO                                                   : Validate the best way to handle options
#   TODO                                                   : Allow specifying the port...?
#   TODO                                                   : Allow for a list of URLs for screenshotting...maybe split this function as a separate function, and the handler calls?  Step function?
    
    configure_binaries()

    chrome_options = webdriver.ChromeOptions()
    chrome_options.add_argument('--headless')
    chrome_options.add_argument("disable-infobars")
    chrome_options.add_argument("enable-automation")
    
    chrome_options.add_argument('--no-sandbox')
    chrome_options.add_argument('--disable-gpu')
    chrome_options.add_argument('--window-size=1280x1696')
    chrome_options.add_argument('--user-data-dir=/tmp/user-data')
    chrome_options.add_argument('--hide-scrollbars')
    chrome_options.add_argument('--enable-logging')
    chrome_options.add_argument('--log-level=0')
    chrome_options.add_argument('--disable-dev-shm-usage')
    chrome_options.add_argument('--v=99')
    chrome_options.add_argument('--single-process')
    chrome_options.add_argument('--data-path=/tmp/data-path')
    chrome_options.add_argument('--ignore-certificate-errors')
    chrome_options.add_argument('--homedir=/tmp')
    chrome_options.add_argument('--disk-cache-dir=/tmp/cache-dir')
    chrome_options.add_argument(
        'user-agent=Mozilla/5.0 (X11; Linux x86_64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/61.0.3163.100 Safari/537.36')
    chrome_options.binary_location = "/tmp/headless-chromium"

    if screenshot_title is None: 
        ext = tldextract.extract(url)
        domain = f"{''.join(ext[:2])}.{ext[2]}:{urlparse(url).port}"
        screenshot_title = f"{domain}_{datetime.utcnow().strftime('%Y%m%d_%H%M%S')}"
    logger.debug(f"Screenshot title: {screenshot_title}")

    with webdriver.Chrome(chrome_options=chrome_options, executable_path="/tmp/chromedriver", service_log_path="/tmp/selenium.log") as driver: # TODO: Is the with loop the best approach?
        driver.set_window_size(1024, 768)
        
        logger.info(f"Obtaining screenshot for {url}")
        driver.get(url)     
        
        sw = driver.execute_script("return document.body.scrollWidth")
        if sw < 1024:
            sw = 1024

        driver.set_window_size(sw, driver.execute_script("return document.body.scrollHeight")+100)

        driver.save_screenshot(f"/tmp/{screenshot_title}.png") # TODO: Delete the screenshot after
        logger.info(f"Uploading /tmp/{screenshot_title}.png to S3 bucket {s3_bucket}/{screenshot_title}.png")
        s3 = boto3.client("s3")
        s3.upload_file(f"/tmp/{screenshot_title}.png", s3_bucket, f"{screenshot_title}.png", ExtraArgs={'ContentType': 'image/png', 'ACL': 'public-read'})
    return f"https://{s3_bucket}.s3.amazonaws.com/{screenshot_title}.png"

def handler(event, context):  
    logger.debug("## ENVIRONMENT VARIABLES ##")
    logger.debug(os.environ)
    logger.debug("## EVENT ##")
    logger.debug(event)

    bucket_name = os.environ["s3_bucket"]
    logger.debug(f"bucket_name: {bucket_name}")

    logger.info("Validating url")  
    if "url" in event: # Using this for testing in lambda.
            try:
                url = event["url"]
            except Exception as e:
                logger.error(e)
                raise e
    else:
        if event["httpMethod"] == "GET":
            if event["queryStringParameters"]:
                try:
                    url = event["queryStringParameters"]["url"]
                except Exception as e:
                    logger.error(e)
                    raise e
            else:
                return {
                    "statusCode": 400,
                    "body": json.dumps("No URL provided...")
                }
        elif event["httpMethod"] == "POST":
            if event["body"]:
                try:
                    body = json.loads(event["body"])
                    url = body["url"]
                except Exception as e:
                    logger.error(e)
                    raise e
            else:
                return {
                    "statusCode": 400,
                    "body": json.dumps("No URL provided...")
                }
        else:
            return {
                "statusCode": 405,
                "body": json.dumps(f"Invalid HTTP Method {event['httpMethod']} supplied")
            }

    logger.info(f"Decoding {url}")
    url = unquote(url)

    logger.info(f"Parsing {url}")
    try: 
        parsed_url = urlparse(url)
        if parsed_url.scheme != "http" and parsed_url.scheme != "https":
            logger.info("No valid scheme found, defaulting to http://")
            parsed_url = urlparse(f"http://{url}")
        if parsed_url.port is None:
            if parsed_url.scheme == "http":
                parsed_url = urlparse(f"{parsed_url.geturl()}:80")
            elif parsed_url.scheme == "https":
                parsed_url = urlparse(f"{parsed_url.geturl()}:443")

    except Exception as e: 
        logger.error(e)
        raise e
    
    logger.info("Getting screenshot")
    try: 
        screenshot_url = get_screenshot(parsed_url.geturl(), bucket_name) # TODO: Variable!
    except Exception as e:  
        logger.error(e)
        raise e

    response_body = {
        "message": f"Successfully captured screenshot of {parsed_url.geturl()}",
        "screenshot_url": screenshot_url
    }

    return {
        "statusCode": 200,
        "body"      : json.dumps(response_body)
    }